### PR TITLE
Fix data corruption in niche use case

### PR DESCRIPTION
**Pull request by terrelln**
* Extract the overflow correction into a helper function.
* Load the dictionary `ZSTD_CHUNKSIZE_MAX = 512 MB` bytes at a time
  and overflow correct between each chunk.
* Add a test case that made overflow asserts fail, and now passes.
   I couldn't figure out how to make a fast test that causes data corruption.

Data corruption could happen when all these conditions are true:

* You are using multithreading mode
* Your overlap size is >= 512 MB (implies window size >= 512 MB)
* You are using a strategy >= ZSTD_btlazy
* You are compressing more than 4 GB

The problem is that when loading a large dictionary we don't do
overflow correction. We can only load 512 MB at a time, and may
need to do overflow correction before each chunk.

Fixes https://github.com/facebook/zstd/issues/1653.

### CODE REVIEW

File: tests/bigdict.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+int main(int argc, const char** argv)
+{
+    ZSTD_CCtx* cctx;
+    ZSTD_DCtx* dctx;
+    const size_t dataSize = (size_t)1 << 30;
+    const size_t outSize = ZSTD_compressBound(dataSize);
+    const size_t bufferSize = (size_t)1 << 31;
+    char* buffer = (char*)malloc(bufferSize);
```
Comment (Cyan4973) [inline tests/bigdict.c]: nit: I guess all these `malloc()` should be `assert()` too ?
